src/clusterfuzz/_internal/bot/tasks/utasks/__init__.py

```diff
@@ -138,6 +140,14 @@ def __exit__(self, _exc_type, _exc_value, _traceback):
     monitoring_metrics.UTASK_SUBTASK_E2E_DURATION_SECS.add(
         e2e_duration_secs, self._labels)
 
+    # The only case where a task might fail without throwing, is in
+    # utask_main, by returning an ErrorType proto which indicates
+    # failure.
+    outcome = 'error' if _exc_type or self.saw_failure else 'success'
+    monitoring_metrics.TASK_OUTCOME_COUNT.increment({
+        **self._labels, 'outcome': outcome
```

Comment:
Since we have a nice enum of possible errors, I'd much rather we be more specific here and record it as the outcome: e.g. `result: 'NO_ERROR'` or `result: 'BUILD_SETUP_FAILED'`. It would greatly help in debugging problems.

Exceptions can be bucketed as their own `result: 'UNHANDLED_EXCEPTION'`.

Do you think it will pose a problem with metric cardinality?

Reply:
Hmm now that I think of it, we should get rid of job

For oss fuzz:

15 tasks
1300 projects ~ 1300 jobs
3 subtasks (pre main post)
2 modes (batch/queue)
3 platforms (linux mac win)
2 outcomes (fail/success)

This goes for around 700k, which will bite us. We could swap job out for all outcomes, and that will be around ~16k possible label combinations, which should be fine. Wdyt?

Reply:
@jonathanmetzman the current state of the subtask duration metric allows for around 350k different label combinations, [GCP's recommendation](https://cloud.google.com/logging/docs/logs-based-metrics/troubleshooting#too-many-time-series) is at most 30k. Should we get rid of job in this context manager as a whole?

Reply:
What's the limit? I think there's other things that can be removed
1. I think modes is definitely not needed, the two modes won't live side by side much longer.
2. I think platform may be unneeded, it can be obtained through the job name.
Would reducing cardinality by 6X help?

Reply:
Also, there's currently about 1000 platforms in oss-fuzz but that is going away. Will that help?

Reply:
It is actually 40 distinct values under that enum.
15 tasks * 40 outcomes * 1300 jobs * 3 subtasks > 1M, so yeah, no chance at all to get the discriminated outcomes, while keeping jobs.

Another option is to spin a separate metric, TASK_OUTCOME_BY_ERROR_TYPE, for which we track all the labels except job, and manage to stay under  30k distinct labels.

Wdyt @letitz

Reply:
Sorry why 40 outcomes?

Reply:
https://github.com/google/clusterfuzz/blob/06173d288ab32f2f1a61e77b083d92b483b00e77/src/clusterfuzz/_internal/protos/uworker_msg.proto#L355

38 ErrorTypes from the utask_main output, 'N/A' (to indicate no error), and 'UNHANDLED_EXCEPTION', as suggested by titouan

Reply:
FIne with me to have a separate metric for error types that does not group by job.

> I think modes is definitely not needed, the two modes won't live side by side much longer.
> I think platform may be unneeded, it can be obtained through the job name.
Would reducing cardinality by 6X help?

Mode does not contribute much to cardinality since we have a mostly static mapping from (task, platform) to mode. Platform is actually free since as you point out, it is implied by job. So there are not 3 copies of each job, one for each platform. IIRC this is what actually matters, not the potential cardinality if we generated all possible labels.